### ISSUE TITLE
JSON Structure

### ISSUE THREAD

**Issue by donmccurdy** [takahirox/KHR_lights_shadows#12]
A few thoughts on JSON structure of the proposed extension —

```json
"nodes": [{
    "extensions": {
        "KHR_lights_punctual": {
            "light": 0,
            "extensions": {
                "KHR_lights_shadows": {
                    "cast": true
                }
            }
        }
    }
}]
```

I would not use nesting here. The `KHR_lights_punctual` object is not a meaningful concept in the scene, it exists only to "extend" the node. Instead think about what would make a clear description if everything were in core...

```jsonc
nodes: [{
  light: 0,
  castShadows: true,
}]
```

... and then extrapolate from that. For example, it's better to say "castShadows" rather than "cast" because the second would have to be changed if it were ever promoted into the core spec. It is similar for material extensions, we say `KHR_materials_iridescence.iridescenceIor` and not `KHR_materials_iridescence.ior`.

**Comment by donmccurdy** [comment]
I am also not so sure about where these definitions should be located, but I would resist the temptation to make those decisions based just on reusability. I could imagine doing this for any material property...

```jsonc
nodes: [{
  mesh: 0,
  baseColorFactor: [1, 1, 1, 1]
}]
```

... and of course it saves a bit of redundant JSON, but ultimately it is making it much more complex for client loaders to know how many materials they're actually going to need to construct in their own runtime representation.

Putting both light-related and mesh- or material-related properties on the node for shadows feels like the same thing.

For the lights, I would be OK with either of these:

```jsonc
// (a)
extensions: {
  "KHR_lights_punctual": {
    "lights": [{
      "extensions": { "KHR_lights_shadows": { castShadows: true }}
    }]
  }
},
nodes: [{
  "extensions": { 
    "KHR_lights_punctual": { "light": 0 }
  },
}],

// (b)
nodes: [{
  "extensions": { 
    "KHR_lights_punctual": { "light": 0 },
    "KHR_lights_shadows": { "castShadows": true }
  },
}],
```

I would exclude (a) from my criticism about nesting above, because a "light" is really a top-level concept in the scene, unlike the extension on a node. But between (a) and (b) I'm not sure, the Khronos folks might have a stronger opinion about nesting.

**Comment by donmccurdy** [comment]
For the properties saying which geometry in the scene casts or receive shadows, valid places to put those might be:

- node
- mesh
- primitive
- material

I believe that decision should probably come down to precedent in existing DCC software and engines, rather than with the goal of minimizing JSON over the wire.

**Comment by takahirox** [comment]
Thanks for the comments! I want to revisit the JSON structure.

**Comment by b1skit** [comment]
Great feedback, makes a lot of sense

**Comment by takahirox** [comment]
@donmccurdy 

> I would not use nesting here. The `KHR_lights_punctual` object is not a meaningful concept in the scene, it exists only to "extend" the node.

Let me confirm my understanding.

From the terminology section in the spec, `node` is described as "An object defining the hierarchy relations and the local transform of its content".

Shadows are scene configuration rather than `node` because cast/receive shadows define whether the shadows of a node object have an effect to/from other nodes. It is beyond the `node` definition. `node` should define just one node and node hierarchy.

`KHR_lights_punctual` is to extend `node`. So shadows configurations shoudln't be in it.

Is my understanding correct? If so, that makes sense to me.

One of the reasons why I thought of nesting in `node` is I thought cast shadows configurations for light should be tight to light. For example it is useless if a `node` has cast shadows for light but doesn't have a `light` like the following. So I thought such useless configuration shouldn't be able to be formed.

```
nodes: [{
  "extensions": { 
    // no lights but shadows configuration for light
    "KHR_lights_shadows": {
      "castShadowsForLight": true
    }
  }
}]
```

Similar for mesh.

```
nodes: [{
  "extensions": { 
    // no mesh but shadows configuration for mesh
    "KHR_lights_shadows": {
      "castShadows": true
    }
  }
}]
```

So, given it and what you mentioned, I started to feel that placing cast/receive shadows configuration under `mesh`/`light` (Option (a) in your examples) may be more natural and straightforward (and easier to handle for extension handlers(?)) ...

> I believe that decision should probably come down to precedent in existing DCC software and engines, rather than with the goal of minimizing JSON over the wire.

Agreed, portability should be prioritized over efficiency because it is one of the core concepts of glTF.

**Comment by b1skit** [comment]
The nesting in (a) seems a bit more intuitive to me.

(b) seems like it might allow the possibility of a node with no light, but a (confusing/meaningless) shadow casting extension property

**Comment by takahirox** [comment]
I moved the extension from under `node` to under `mesh` and `light`, and also renamed properties from `cast`/`receive` to `castShadows/receiveShadows` in #13 #14

> For the properties saying which geometry in the scene casts or receive shadows, valid places to put those might be:
>
>node
>mesh
>primitive
>material

`primitive` instead of `mesh` was another option in my mind. But it may not fit if glTF loader creates a mesh object, not multiple mesh objects, from a `mesh` definition having multiple primitives as Three.js used to do. So I chose `mesh` for now.

> I believe that decision should probably come down to precedent in existing DCC software and engines

I want to propose the extension to Khronos group and discuss more details there because checking all major existing software by myself is hard.
